### PR TITLE
Dixon (dense) refacto

### PR DESCRIPTION
**Pull request by Breush**
- Noticed issue #197

### CODE REVIEW

File: linbox/algorithms/dixon-solver/dixon-solver-common.inl [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	{
+		return p < Prime(67108863);
+	}
+
+	template<>
+	inline bool checkBlasPrime(const BlasVector<Givaro::ZRing<Integer>>& p)
+	{
+		bool tmp=true;
+		for (size_t i=0;i<p.size();++i)
+			if  (p[i] >= integer(67108863)) {
```
Comment (jgdumas) [inline linbox/algorithms/dixon-solver/dixon-solver-common.inl]: At least call the above checkBlasPrime

File: linbox/algorithms/dixon-solver/dixon-solver-dense.inl [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+    template <class Ring, class Field, class RandomPrime>
+    template <class IMatrix, class Vector1, class Vector2>
+    SolverReturnStatus DixonSolver<Ring, Field, RandomPrime, Method::DenseElimination>::findRandomSolution(
+        Vector1& num, Integer& den, const IMatrix& A, const Vector2& b, int maxPrimes, const SolverLevel level)
+    {
+
+        return monolithicSolve(num, den, A, b, false, true, maxPrimes, level);
+    }
+
+    // TAS stands for Transpose Augmented System (A|b)t
```
Comment (jgdumas) [inline linbox/algorithms/dixon-solver/dixon-solver-dense.inl]: Call the class TransposeAugmentedSystem ?

File: linbox/algorithms/rational-solver2.h [hunk 3]
```diff
@@ -56,15 +56,15 @@ namespace LinBox
 	//template argument Field and RandomPrime are not used.
 	//Keep it just for interface consistency.
 	template <class Ring, class Field, class RandomPrime>
-	class RationalSolver<Ring, Field, RandomPrime, Method::SymbolicNumericNorm> {
+	class DixonSolver<Ring, Field, RandomPrime, Method::SymbolicNumericNorm> {
```
Comment (jgdumas) [inline linbox/algorithms/rational-solver2.h]: Call the file symbolic-numeric-dixon-solve ?

File: tests/test-solve-full.C [hunk 4]
```diff
@@ -262,89 +262,89 @@ int main(int argc, char** argv)
...
+        // // @deprecated These do not compile anymore
+        // // ok = ok && test_dense_solve(Method::BlockWiedemann(method), F, F, m, n, 0, 0, seed, verbose);
+        // // ok = ok && test_sparse_solve(Method::BlockWiedemann(method), F, F, m, n, 0, 0, seed, verbose);
+        // // ok = ok && test_blackbox_solve(Method::BlockWiedemann(method), F, F, m, n, 0, 0, seed, verbose);
+
+        // // ----- Modular Coppersmith
+        // // @deprecated These do not compile anymore
+        // // ok = ok && test_dense_solve(Method::Coppersmith(method), F, F, m, n, 0, 0, seed, verbose);
+        // // ok = ok && test_sparse_solve(Method::Coppersmith(method), F, F, m, n, 0, 0, seed, verbose);
+        // // ok = ok && test_blackbox_solve(Method::Coppersmith(method), F, F, m, n, 0, 0, seed, verbose);
```
Comment (jgdumas) [inline tests/test-solve-full.C]: Hum, what remains uncommented ?

**Comment by Breush** [comment]
@ClementPernet In the `class TransposeAugmentedSystem`, we extract permutations of the augmented system (A|b)t = PLUQ into indices `srcRow, srcCol`. Does FFLAS has something for this?

File: linbox/algorithms/dixon-solver/dixon-solver-dense.inl [hunk 5]
```diff
@@ -215,11 +210,13 @@ namespace LinBox {
             P.resize(factors->rowdim());
             Q.resize(factors->coldim());
 
+            std::cout << "factors: " << *factors << std::endl;
```
Comment (jgdumas) [inline linbox/algorithms/dixon-solver/dixon-solver-dense.inl]: protect comments

**Comment by jgdumas** [comment]
Pb

g++ -DHAVE_CONFIG_H -I. -I..  -DLinBoxTestOnly -O2 -Wall -g -DDEBUG -D_LB_DEBUG -g -DDEBUG -D_LB_DEBUG -I.. -fopenmp -fabi-version=6 -I/usr/local/soft/givaro-4.0/include -I/data/scratch/gmp-6.1.1//include -I/usr/local/soft/fflas-ffpack-2.3/include -I/data/scratch/ntl-11.3.2/include       -msse -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mavx -MT test-det.o -MD -MP -MF .deps/test-det.Tpo -c -o test-det.o test-det.C
In file included from ../linbox/algorithms/hybrid-det.h:36,
                 from ../linbox/solutions/det.h:600,
                 from test-det.C:48:
../linbox/algorithms/last-invariant-factor.h: In instantiation of ‘LinBox::LastInvariantFactor<_Ring, _Solver>::Integer& LinBox::LastInvariantFactor<_Ring, _Solver>::lastInvariantFactor1(LinBox::LastInvariantFactor<_Ring, _Solver>::Integer&, Vector&, const IMatrix&, bool) const [with IMatrix = LinBox::SparseMatrix<Givaro::IntegerDom>; Vector = LinBox::BlasVector<Givaro::IntegerDom>; _Ring = Givaro::IntegerDom; _Solver = LinBox::DixonSolver<Givaro::IntegerDom, Givaro::ModularBalanced<double>, LinBox::PrimeIterator<LinBox::IteratorCategories::HeuristicTag>, LinBox::Method::DenseElimination>; LinBox::LastInvariantFactor<_Ring, _Solver>::Integer = Givaro::Integer]’:
../linbox/algorithms/hybrid-det.h:267:47:   required from ‘typename Blackbox::Field::Element& LinBox::lif_cra_det(typename Blackbox::Field::Element&, const Blackbox&, const LinBox::RingCategories::IntegerTag&, const MyMethod&) [with Blackbox = LinBox::SparseMatrix<Givaro::IntegerDom>; MyMethod = LinBox::Method::Wiedemann; typename Blackbox::Field::Element = Givaro::Integer]’
../linbox/solutions/det.h:620:26:   required from ‘typename Blackbox::Field::Element& LinBox::det(typename Blackbox::Field::Element&, const Blackbox&, const LinBox::RingCategories::IntegerTag&, const MyMethod&) [with Blackbox = LinBox::SparseMatrix<Givaro::IntegerDom>; MyMethod = LinBox::Method::Wiedemann; typename Blackbox::Field::Element = Givaro::Integer]’
../linbox/solutions/det.h:107:13:   required from ‘typename Blackbox::Field::Element& LinBox::det(typename Blackbox::Field::Element&, const Blackbox&, const MyMethod&) [with Blackbox = LinBox::SparseMatrix<Givaro::IntegerDom>; MyMethod = LinBox::Method::Wiedemann; typename Blackbox::Field::Element = Givaro::Integer]’
test-det.C:368:49:   required from here
../linbox/algorithms/last-invariant-factor.h:293:8: error: passing ‘const Solver’ {aka ‘const LinBox::DixonSolver<Givaro::IntegerDom, Givaro::ModularBalanced<double>, LinBox::PrimeIterator<LinBox::IteratorCategories::HeuristicTag>, LinBox::Method::DenseElimination>’} as ‘this’ argument discards qualifiers [-fpermissive]
    tmp = solver.solveNonsingular(r_num, r_den, A, b,oldMatrix);
In file included from ../linbox/algorithms/./dixon-solver/dixon-solver-dense.h:363,
                 from ../linbox/algorithms/rational-solver.h:583,
                 from ../linbox/algorithms/hybrid-det.h:35,
                 from ../linbox/solutions/det.h:600,
                 from test-det.C:48:
../linbox/algorithms/./dixon-solver/./dixon-solver-dense.inl:63:24: note:   in call to ‘LinBox::SolverReturnStatus LinBox::DixonSolver<Ring, Field, RandomPrime, LinBox::Method::DenseElimination>::solveNonsingular(Vector1&, LinBox::DixonSolver<Ring, Field, RandomPrime, LinBox::Method::DenseElimination>::Integer&, const IMatrix&, const Vector2&, bool, int) [with IMatrix = LinBox::SparseMatrix<Givaro::IntegerDom>; Vector1 = LinBox::BlasVector<Givaro::IntegerDom>; Vector2 = LinBox::BlasVector<Givaro::IntegerDom>; Ring = Givaro::IntegerDom; Field = Givaro::ModularBalanced<double>; RandomPrime = LinBox::PrimeIterator<LinBox::IteratorCategories::HeuristicTag>; LinBox::DixonSolver<Ring, Field, RandomPrime, LinBox::Method::DenseElimination>::Integer = Givaro::Integer]’
     SolverReturnStatus DixonSolver<Ring, Field, RandomPrime, Method::DenseElimination>::solveNonsingular(
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../linbox/algorithms/hybrid-det.h:36,
                 from ../linbox/solutions/det.h:600,
                 from test-det.C:48:
../linbox/algorithms/last-invariant-factor.h: In instantiation of ‘LinBox::LastInvariantFactor<_Ring, _Solver>::Integer& LinBox::LastInvariantFactor<_Ring, _Solver>::lastInvariantFactor1(LinBox::LastInvariantFactor<_Ring, _Solver>::Integer&, Vector&, const IMatrix&, bool) const [with IMatrix = LinBox::BlasMatrix<Givaro::IntegerDom>; Vector = LinBox::BlasVector<Givaro::IntegerDom>; _Ring = Givaro::IntegerDom; _Solver = LinBox::DixonSolver<Givaro::IntegerDom, Givaro::Modular<double>, LinBox::PrimeIterator<LinBox::IteratorCategories::HeuristicTag>, LinBox::Method::DenseElimination>; LinBox::LastInvariantFactor<_Ring, _Solver>::Integer = Givaro::Integer]’:
../linbox/algorithms/det-rational.h:251:32:   required from ‘typename Rationals::Element& LinBox::rational_det(typename Rationals::Element&, const LinBox::BlasMatrix<_Field>&, const MyMethod&) [with Rationals = Givaro::QField<Givaro::Rational>; MyMethod = LinBox::Method::Auto; typename Rationals::Element = Givaro::Rational; typename LinBox::Vector<Ring>::Dense = std::vector<Givaro::Rational, std::allocator<Givaro::Rational> >]’
../linbox/solutions/det.h:660:15:   required from ‘typename Field::Element& LinBox::det(typename Field::Element&, const LinBox::BlasMatrix<_Field>&, const LinBox::RingCategories::RationalTag&, const MyMethod&) [with Field = Givaro::QField<Givaro::Rational>; MyMethod = LinBox::Method::Auto; typename Field::Element = Givaro::Rational; typename LinBox::Vector<Ring>::Dense = std::vector<Givaro::Rational, std::allocator<Givaro::Rational> >]’
../linbox/solutions/det.h:107:13:   required from ‘typename Blackbox::Field::Element& LinBox::det(typename Blackbox::Field::Element&, const Blackbox&, const MyMethod&) [with Blackbox = LinBox::BlasMatrix<Givaro::QField<Givaro::Rational>, std::vector<Givaro::Rational, std::allocator<Givaro::Rational> > >; MyMethod = LinBox::Method::Auto; typename Blackbox::Field::Element = Givaro::Rational]’
test-det.C:566:41:   required from here
../linbox/algorithms/last-invariant-factor.h:293:8: error: passing ‘const Solver’ {aka ‘const LinBox::DixonSolver<Givaro::IntegerDom, Givaro::Modular<double>, LinBox::PrimeIterator<LinBox::IteratorCategories::HeuristicTag>, LinBox::Method::DenseElimination>’} as ‘this’ argument discards qualifiers [-fpermissive]
    tmp = solver.solveNonsingular(r_num, r_den, A, b,oldMatrix);
In file included from ../linbox/algorithms/./dixon-solver/dixon-solver-dense.h:363,
                 from ../linbox/algorithms/rational-solver.h:583,
                 from ../linbox/algorithms/hybrid-det.h:35,
                 from ../linbox/solutions/det.h:600,
                 from test-det.C:48:
../linbox/algorithms/./dixon-solver/./dixon-solver-dense.inl:63:24: note:   in call to ‘LinBox::SolverReturnStatus LinBox::DixonSolver<Ring, Field, RandomPrime, LinBox::Method::DenseElimination>::solveNonsingular(Vector1&, LinBox::DixonSolver<Ring, Field, RandomPrime, LinBox::Method::DenseElimination>::Integer&, const IMatrix&, const Vector2&, bool, int) [with IMatrix = LinBox::BlasMatrix<Givaro::IntegerDom>; Vector1 = LinBox::BlasVector<Givaro::IntegerDom>; Vector2 = LinBox::BlasVector<Givaro::IntegerDom>; Ring = Givaro::IntegerDom; Field = Givaro::Modular<double>; RandomPrime = LinBox::PrimeIterator<LinBox::IteratorCategories::HeuristicTag>; LinBox::DixonSolver<Ring, Field, RandomPrime, LinBox::Method::DenseElimination>::Integer = Givaro::Integer]’
     SolverReturnStatus DixonSolver<Ring, Field, RandomPrime, Method::DenseElimination>::solveNonsingular(
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[4]: *** [Makefile:2583: test-det.o] Error 1

**Pull request merged**